File: Algorithms/HeapSort/code/src/HeapSort.c
#include "../inc/HeapSort.h"

void heapSort(int *arr, int len) {
    for(int i = len/2 - 1; i > -1; i--) {
        heapAdjust(arr, i, len);
    }
}

void heapAdjust(int *arr, int k, int len) {
    int temp = arr[k];
    for(int i = (k+1)*2 - 1; i < len; i = (i+1)*2 - 1) {
        if(i+1 < len && arr[i+1] > arr[i]) {
            i++;
        }
        if(arr[i] > temp) {
            arr[k] = arr[i];
            k = i;
        } else {
            break;
        }
        arr[k] = temp;
    }
}

File: Algorithms/HeapSort/code/heapSort.py

def heapSort(arr: list):

    def adjust(k: int):
        j = (k+1)*2 - 1
        while j < len(arr):
            if j+1 < len(arr) and arr[j+1] > arr[j]:
                j += 1
            if arr[k] > arr[j]:
                break
            arr[k], arr[j] = arr[j], arr[k]
            k = j
            j = (k+1)*2 - 1
            pass

    for i in range(int(len(arr)/2) - 1, -1, -1):
        adjust(i)


if __name__ == "__main__":
    array = [1, 4, 6, 7, 2, 34, 56, 12, 53, 75, ]
    heapSort(array)
    print(array)
